firestore/integration_test_internal/src/validation_test.cc

```diff
@@ -19,27 +19,276 @@
-// we do not throw exception while only dump exception info to logs. However, in
-// order to test this behavior, we enable exception here and check exceptions.
+// PORT_NOTE: C++ exceptions are forbidden by the Google style guide. However,
+// when building the Unity SDK, we build C++ with exceptions enabled which
+// allows us to propagate C++ exceptions as C# exceptions. The exact exceptions
+// being thrown are easier to check on the C++ layer, so these tests must be
+// built with exceptions enabled.
 
 namespace firebase {
 namespace firestore {
 
-// This eventually works for iOS as well and becomes the cross-platform test for
-// C++ client SDK. For now, only enabled for Android platform.
-
 #if FIRESTORE_HAVE_EXCEPTIONS
 
 using testing::AnyOf;
+using testing::HasSubstr;
 using testing::Property;
 using testing::StrEq;
 using testing::Throws;
 
-#define EXPECT_ERROR(stmt, message)                           \
+#define EXPECT_ERROR(stmt, msg)                               \
   EXPECT_THAT([&] { stmt; }, Throws<std::exception>(Property( \
-                                 &std::exception::what, StrEq(message))));
+                                 &std::exception::what, StrEq(msg))));
+
+#define EXPECT_ERROR_EITHER(stmt, msg1, msg2)  \
+  EXPECT_THAT([&] { stmt; },                   \
+              Throws<std::exception>(Property( \
+                  &std::exception::what, AnyOf(StrEq(msg1), StrEq(msg2)))));
+
+namespace {
+
+enum class ErrorCase {
+  kSettingsAfterUse,
+  kSettingsDisableSsl,
+  kFieldValueDeleteInSet,
+  kFieldValueDeleteNested,
+  kFieldNameEmpty1,
+  kFieldNameEmpty2,
+  kArrayUnionInQuery,
+  kArrayRemoveInQuery,
+  kQueryMissingOrderBy,
+  kQueryOrderByTooManyArguments,
+  kQueryInvalidBoundInteger,
+  kQueryInvalidBoundWithSlash,
+  kQueryDifferentInequalityFields,
+  kQueryInequalityOrderByDifferentFields,
+  kQueryMultipleArrayContains,
+  kQueryStartBoundWithoutOrderBy,
+  kQueryEndBoundWithoutOrderBy,
+  kQueryDocumentIdEmpty,
+  kQueryDocumentIdSlash,
+  kQueryDocumentIdInteger,
+  kQueryDocumentIdArrayContains
+};
+
+// Returns the exact error message used on a platform (in many cases, Android is
+// different from other platforms because it wraps the actual Android SDK).
+// TODO(b/171990785): Unify Android and C++ validation error messages.
+std::string ErrorMessage(ErrorCase error_case) {
+  switch (error_case) {
+    case ErrorCase::kSettingsAfterUse:
+#ifdef __ANDROID__
+      return "FirebaseFirestore has already been started and its settings can "
+             "no longer be changed. You can only call setFirestoreSettings() "
+             "before calling any other methods on a FirebaseFirestore object.";
+#else
+      return "Firestore instance has already been started and its settings can "
+             "no longer be changed. You can only set settings before calling "
+             "any other methods on a Firestore instance.";
+#endif
+
+    case ErrorCase::kSettingsDisableSsl:
```

Comment:
The two values for `kSettingsDisableSsl` are identical. Can you get rid of the `#ifdef` for this one?

Reply:
Thanks for spotting this! I think I originally thought that the C++ validation message would spell it `ssl_enabled`, but it turned out to be not true.